### DISCUSSION TITLE
feature request: host keyword

### DISCUSSION THREAD

**Discussion by xoid** [PerlDancer/Dancer2#1699] [Ideas]
Would be great to have host predicate to serve multiple domainnames:
get '/' => host 'my.host.com' => sub {do something};

**Comment by GeekRuthie** [comment]
I like this idea @xoid!  I actually have a use case where I've got a site working off of a domain, and a subdomain of itself. The difference is **very** small, so I handle it with a 'before' hook to set a var if we're in the subdomain and just check for that var in the routes.  But I can see the use case where you had radically different code, where a predicate like this would be useful; it would avoid potentially-large if-blocks in your code, which appeals to my sense of cleanliness, too.

I have an idea how to do this as a plugin, and think I could have that ready for you to use later this week or this weekend.  It would take a string or a regex, so you could do:
```

`get '/' => host 'thishost.thatsite.com' => sub {  };    # very specific route
get '/' => host qr/.*\.thissite\.com/ => sub { };         # somewhat specific route
get '/' => host qr/.*/ => sub { };                                # default route
get '/' => host 'default' => sub {};                            # synonym for qr/.*/
```
``

`Would something like that work for your use-case?

**Reply by xoid** [reply]
Yes, @GeekRuthie!
I've been desperately trying for weeks to write Dancer2::Plugin::Host taking as an example the predicates "need login" and "at thursday" from Plugin::Manual and from the Advent calendar articles. It is confusing that some articles describe a standalone dancer server (*.pl), others "PSGI" or an application created through the CLI generator. In general, I was too lame to understand core logic. And I am very grateful that you came to help. Thanks! I would also like to offer the option to divide host logic to different files/modules: 

    get '/' => host 'my.host.com' => MyHost::get_handler()
    get '/' => host 'another.host.com' => AnotherHost::get_handler()

or something like that. (It looks ugly yet)
Indeed, anonymous function are very confusing as bad as anonymous variables and I avoid them if possible. Names are speaking for itself! Also anonymous functions tend to ugly (for newbie like me) code formating

**Reply by GeekRuthie** [reply]
Referring to other modules should actually work fine already; the value of a route or predicate hashref is supposed to be a coderef, so if your subroutines are properly exported and visible to the route code, that should work without any trouble.  I'll definitely test for that, though!

I'll try to get on that this week; I'll probably call it Dancer2::Plugin::HostSpecificRoute, just so it's super-clear what it's supposed to do.

Thanks again for the great idea!

**Comment by GeekRuthie** [comment]
Success!

I've posted the new plugin you suggested to PAUSE, and it should be available shortly on [CPAN](https://metacpan.org/pod/Dancer2::Plugin::HostSpecificRoute).  

This was a great idea, and fortunately not at all difficult to deliver.  Please let me know via the Issues tracker on the repository if you find any problems.  Thanks for being a part of our community, @xoid !
